### PR TITLE
test: replace Karma with Webdriver.IO

### PR DESCRIPTION
**Pull request by christian-bromann**
The current test framework for browser testing (Karma) is not maintained anymore and WebdriverIO provides a more modern stack that allows to test in different browser.

This patch replaces these test frameworks.

fixes: #17009

#### Prerequisites checklist

- [ ] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[x] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

#### Is there anything you'd like reviewers to focus on?

This is still a work in progress PR but I would like the maintainer to take a quick look and verify that they are ok with the change.

### CODE REVIEW

File: lib/linter/rules.js
Comment (aladdin-add) [inline lib/linter/rules.js:59]: Can you explain why changing it? a rule may be loaded from a third-party plugin, so this won't always work.

Comment (christian-bromann) [reply lib/linter/rules.js:59]: I see. Transpiling dynamic CJS imports into ESM is kinda of tricky so this was for now the only way to workaround it. I was able to verify that this code is not touched during the tests so it shouldn't be modified. I will work on reverting it.

File: package.json
Comment (aladdin-add) [inline package.json:155]: question: do we need switch puppeteer to something like playwright too? 😄

**Comment by mdjermanovic** [comment]
I converted this to draft to prevent accidental merging since it's still work in progress and some tests are `.skip`-ed,

File: package.json
Comment (christian-bromann) [reply package.json:155]: No, WebdriverIO does all the browser automation. I wonder if having Puppeteer here is still needed. I will check.

**Comment by christian-bromann** [comment]
> Can we still use webpack

WebdriverIO uses Vite to power the component tests (including mocking etc.). Let me check if I can just bundle with webpack and import the compiled files. This would make the setup also simpler.

**Comment by Rec0iL99** [comment]
Don't close. Work in progress.

**Comment by christian-bromann** [comment]
Sorry, haven't forgotten about this. Will try to put some time in this or next week to finish it.

**Comment by aladdin-add** [comment]
@christian-bromann no worries, I've removed it.

**Comment by christian-bromann** [comment]
Made some progress on this:

- removed the changes in EsLint core
- use the webpack build artifact within the browser test
- removed unused puppeteer dependency
- used WebdriverIO based mocking for two tests as Sinon can't spy on ESM modules

Note: for the mocking to work I had to use import syntax. I will have to push an enhancement to WebdriverIO to switch back to require but I wonder if this test is also run in Node.js because if not I can just make everything an import statement.

**Comment by christian-bromann** [comment]
The build is failing due to:

```
/home/runner/work/eslint/eslint/tests/lib/linter/linter.js
  12:1  error  Parsing error: 'import' and 'export' may appear only with 'sourceType: module'
```

Unfortunately WebdriverIOs mocking features only work when using `import`. Since the `linter.js` test is run within the browser I wonder if we can move the file into a directory for browser tests and change the file to use ESM. What do you think?

**Comment by mdjermanovic** [comment]
> Note: for the mocking to work I had to use import syntax. I will have to push an enhancement to WebdriverIO to switch back to require but I wonder if this test is also run in Node.js because if not I can just make everything an import statement

This test file is run in Node.js too. Perhaps we could make another bundle from `tests/lib/linter/linter.js` entry point and then run that file in a browser?

**Comment by nzakas** [comment]
@christian-bromann are you still working on this?

**Comment by christian-bromann** [comment]
@nzakas yes, sorry for my delay, still many competing tasks but will try to get back to it in the following days

**Comment by BYK** [comment]
@christian-bromann do you need any help with pushing this over the finish line? Looks like it is close so I'd like to help if you need any.

**Comment by christian-bromann** [comment]
Thanks for pinging and sorry for the delay. I have been 2 weeks on vacation. Put it on my todo for this week.

**Comment by christian-bromann** [comment]
@nzakas @BYK this is ready for review now

File: tests/lib/linter/linter.js
Comment (kecrily) [inline tests/lib/linter/linter.js:1]: Should be rename to `linter.mjs`

Comment (christian-bromann) [reply tests/lib/linter/linter.js:1]: Thanks for the review, made the changes

**Comment by mdjermanovic** [comment]
We're running `mocha` on `.js` files only, so after switching the linter test file to `.mjs`, those tests won't be run in Node.

https://github.com/eslint/eslint/blob/c6f8cd0d62e4a3c314c6860ff367490bbd05325a/Makefile.js#L79

Why was it necessary to convert the file to ESM?

**Comment by kecrily** [comment]
Because of linter and test errors

See https://github.com/eslint/eslint/actions/runs/5482202414/jobs/10005549763?pr=17126 and https://github.com/eslint/eslint/actions/runs/5482202414/jobs/10005549380?pr=17126

**Comment by kecrily** [comment]
I also found missing compatibility code to use both import and require in the same file.

```
import { createRequire } from "module";
const require = createRequire(import.meta.url);

```

But when I added it locally and tested it, I got some test errors. Need to investigate further

image

**Comment by mdjermanovic** [comment]
Why do we need to `import` modules? That seems like an extensive change in a test file whose primary purpose is to test how the core linter module works in Node.js.

**Comment by christian-bromann** [comment]
> Why do we need to `import` modules? That seems like an extensive change in a test file whose primary purpose is to test how the core linter module works in Node.js.

These tests are run in the browser. It is currently a limitation of WebdriverIO. I didn't know these tests were also run in Mocha which makes everything a bit more complicated. I will see what I can do.

**Comment by mdjermanovic** [comment]
Could we maybe simply webpack `tests/lib/linter/linter.js` and load the bundle in the browser as `specs`?

Here I made a [test repo](https://github.com/mdjermanovic/linter-browser-test) with the bundle (`linter.test.js`) where you can run it locally (`npm run mocha`), in browser with karma (`npm run karma`), or directly in browser (open https://mdjermanovic.github.io/linter-browser-test/).

**Comment by nzakas** [comment]
Agree with others, we should not be changing `linter.js` to ESM -- we have a lot of changes coming and going in there, and this would complicate testing without providing much additional value.

And yes, we need all of the tests to stay in `linter.js` to make it easier as we implement the new config system.

**Comment by christian-bromann** [comment]
Thank you all for your feedback!

I made it work 🎉  with the most minimal changes to the `linter.js` file as possible. I wrote a custom Vite plugin that would help me to dynamically update the import from the `linter.js` in `src` to `build`. I had to skip 2 tests because the way they spy on the module is not possible in the browser. I am not sure how Webpack was configured in Karma to make this possible but I think this is fine for now.

Let me know what you think!

File: wdio.conf.js
Comment (aladdin-add) [inline wdio.conf.js:113]: nits: I prefer something like `"./wdio-logs"`, to avoid conflicting with other tools. :)

File: wdio.conf.js
Comment (aladdin-add) [inline wdio.conf.js:22]: not sure this is needed.  IMHO, the webpack bundled file should have included node.js built-in modules, as configured: https://github.com/eslint/eslint/blob/8bcbf11b6050418262ffa8e0ca37f365ae92e7ce/webpack.config.js#L47

Comment (christian-bromann) [reply wdio.conf.js:22]: WebdriverIOs component testing framework uses Vite under the hood, which allows Rollup plugins.

Comment (aladdin-add) [reply wdio.conf.js:22]: well, I see!

the plugin was applied to `/lib/linter/linter.js`, but not `tests/lib/linter/linter.js`. 😄

**Review by aladdin-add — APPROVED** [review]
LGTM, thanks!

**Review by aladdin-add — CHANGES_REQUESTED** [review]
the ci was failing as some dev deps were not compat with node.js v12.( it has been EOL, but eslint v8 still supports.)

I think we can merge it when we drop node.js v12 support.

**Comment by kecrily** [comment]
But this is a devDependencies change, and I don't think it will break ESLint's support for node 12.

**Comment by aladdin-add** [comment]
yes, but it causes ci to fail and we are not able to merge it.

**Comment by christian-bromann** [comment]
> the ci was failing as some dev deps were not compat with node.js v12.( it has been EOL, but eslint v8 still supports.)

The [migration docs](https://eslint.org/docs/latest/use/migrate-to-8.0.0) state that v8 _"is officially dropping support for [Node.js 10, 13, 15] starting with ESLint v8.0.0"_. Am I missing something?

File: package.json
Comment (aladdin-add) [inline package.json:114]:
```suggestion
    "chromedriver": "^106.0.0",
```

it should be fixed if back to the version: https://github.com/aladdin-add/eslint/actions/runs/5571225831/jobs/10176049189?pr=21

**Comment by christian-bromann** [comment]
@aladdin-add I reverted the version update of Chromedriver and added flag so that it detects the right version in CI. However this could confuse contributor trying to run the test and don't have the right Chromedriver version installed. I am working to get this all embedded into WebdriverIO so no one has to worry about the driver.

**Comment by christian-bromann** [comment]
I am working on some WebdriverIO things to get the issue in the browser test elevated.

**Comment by christian-bromann** [comment]
> I'm curious, why the problem with spying on ES modules appears here, since this test doesn't seem to use any ES modules?

These tests run in the browser so they are all ES modules. I am not sure how WebPack transpiled the code to not have this error thrown but basically all imported modules in the browser are sealed and can't be changed. However I pushed a commit that changes the tests so that it should work in both environments. Also I made some changes to WebdriverIO to cope some performance issues. I think it should all pass now 🤞

**Review by aladdin-add — APPROVED** [review]
✅all green, thanks!

just waiting for @mdjermanovic and other members to take another look before merging.

**Review by nzakas — APPROVED** [review]
LGTM. Thanks for all of your work on this.

Just waiting for @mdjermanovic to re-review.

File: .github/workflows/ci.yml
Comment (snitin315) [inline .github/workflows/ci.yml:88]: How will we download & see these logs?

Comment (christian-bromann) [reply .github/workflows/ci.yml:88]: > How will we download & see these logs?

In the summary of a job under `Artifacts`, see picture below (bottom):
Screenshot 2023-07-26 at 21 27 09

https://github.com/eslint/eslint/actions/runs/5583652074

**Comment by mdjermanovic** [comment]
I'm getting this locally:

```
C:\milos\tmp\wsdio-eslint>npm run wdio

> eslint@8.45.0 wdio
> wdio run ./wdio.conf.js


Execution of 1 workers started at 2023-07-27T12:26:28.122Z

[0-0] RUNNING in chrome - file:///C:/milos/tmp/wsdio-eslint/tests/lib/linter/linter.js

DevTools listening on ws://127.0.0.1:64612/devtools/browser/99c26fcd-2ca5-48b8-8dcd-a5fcdd775b26

DevTools listening on ws://127.0.0.1:64618/devtools/browser/f38249c5-2e5e-4959-812e-a00d2f3fc2d4

DevTools listening on ws://127.0.0.1:64624/devtools/browser/33a02012-1c10-45e4-a678-6ee2d6a551cd

DevTools listening on ws://127.0.0.1:64630/devtools/browser/9b25ff2d-d847-41b5-a43a-b195ac9ca8d6
[0-0] FAILED in chrome - file:///C:/milos/tmp/wsdio-eslint/tests/lib/linter/linter.js

 "concise" Reporter:
========= Your concise report ==========
undefined
❌ Failed to setup tests, no tests found


Spec Files:      0 passed, 1 failed, 1 total (100% completed) in 00:00:06
```

**Comment by christian-bromann** [comment]
@kecrily @mdjermanovic I updated the branch, all wdio dependencies and made the path to the spec file more explicit. This should resolve the issue for you (I assume you've been running the command outside of the root dir). Let me know if there is anything else I can do.

**Comment by nzakas** [comment]
@christian-bromann the test is now failing on the PR

**Comment by christian-bromann** [comment]
@nzakas pushed a fix for it. I missed to remove the Chromedriver dependencies which is now all built-into WebdriverIO. Due to differences between CI and my local machine it passed when I tested it but failed in CI. Should be good to go now.

File: package.json
Comment (kecrily) [inline package.json:33]: It doesn't match [our Package.json Conventions](https://github.com/eslint/eslint/blob/main/docs/src/contribute/package-json-conventions.md), can you modify it?

Comment (christian-bromann) [reply package.json]: ahh, good catch 👍  removed.

**Comment by aladdin-add** [comment]
seems working fine on my macos! 👍

**Review by kecrily — APPROVED** [review]
LGTM

**Review by kecrily — APPROVED** [review]
LGTM

**Comment by aladdin-add** [comment]
merged, thanks for contributing! 👍

**Comment by christian-bromann** [comment]
Thanks folks for the support 👍 don't hesitate to ping me if you run into any issues with the tests.

**Comment by mdjermanovic** [comment]
Question: Does this expect that Chrome is already installed and what happens if it isn't? Asking because we previously had issues with the browser test on our Jenkins server that we use for releases, and then we added `puppeteer` to fix the problem (https://github.com/eslint/eslint/pull/11027). But, `puppeteer` is removed now, so I'm worried if this test will work on our Jenkins server.

**Comment by christian-bromann** [comment]
Currently WebdriverIO will prefer the browser installed on the host OS, if no Chrome is installed it will automatically install latest stable release of Chrome. If you set `browserVersion: 'stable'` in your capabilities it will always download the latest stable release.

**Pull request merged**
